### ISSUE TITLE
Semaphores Unit Test: Cannot Create Semaphores

### ISSUE THREAD

**Issue by Theldus** [nanvix/nanvix#222]
**Description**

While running `test sem` the semaphores cannot be allocated and the following error message appears: cannot create semaphores. 

This problem occurs because of the way the parameters are passed to the `sys_semopen` function: two parameters and two _var_args_. It turns out that var_args are passed differently than i386, and when the 'mode' variable is retrieved with something like `va_arg (arg, mode_t)` the following scenario occurs:

```
i386                       or1k
stack                      stack
 -----                      -----
|value|                    |     |
|mode |                    |     |
|oflag| <- va_start        |     |
|name |                    |     | <- va_start
 -----                      -----
                            value  <- r6 |\
                            mode   <- r5 | \
                            oflag  <- r4 | / -> registers
                            name   <- r3 |/
```
The ABI of or1k specifies that up to six parameters must be passed via registers (r3-r8) and the remaining via stack. Variable arguments follow the same idea, so only the additional parameters will be passed via stack and the remaining via registers. Therefore, the ideal layout for the correct operation would be:
```
or1k
stack
 -----
|value|
|mode | <- va_start
 ----- 
 oflag  <- r4
 name   <- r3
```
While this works very well in user space, this requires the kernel to know in advance which parameter starts the list of additional parameters, so it pushes them. It's important to note that stacking all does not solve the problem, since it expects the top of the stack to be the first var_arg element, not the first parameter.

**Solution**

Perhaps the best solution is to not use var_args between the syscall entry-point and handlers, since the entry-point is unaware of the use of var_args. This only works on i386 because the parameters are always passed via stack.

**Comment by Theldus** [comment]
I've made a push of a branch called 'fix-semaphores-222', with a proposed solution, if it's good enough, I can create a PR from it.

**Comment by ppenna** [comment]
@Theldus I totally approve this solution. Commit 4e2d77b3dfed9fe18e183ca328e75ea0e049e2db provides a long-term solution for this kind of problem.

Indeed, we should avoid using var lists on the kernel. As you have noticed they are kind architecture dependent. =)

Go ahead and open a pull request ;-)

**Comment by ppenna** [comment]
Issue fixed on commit 4e2d77b3dfed9fe18e183ca328e75ea0e049e2db.
